pubsub/producer.go

```diff
@@ -102,150 +92,131 @@ func NewProducer[Request any, Response any](client redis.UniversalClient, stream
 func (p *Producer[Request, Response]) checkResponses(ctx context.Context) time.Duration {
-	minIdInt := [2]uint64{math.MaxUint64, math.MaxUint64}
+	pelData, err := p.client.XPending(ctx, p.redisStream, p.redisGroup).Result()
+	if err != nil {
+		log.Error("error getting PEL data from xpending, xtrimming is disabled", "err", err)
+	}
+	deletePromise := func(id string) {
+		// Try deleting UNIQUEID_MSGID_MAP_KEY corresponding to this id from redis
+		if err := p.client.Del(ctx, MessageKeyFor(p.redisStream, id)+UNIQUEID_MSGID_MAP_KEY).Err(); err != nil {
+			log.Error("Error deleting key from redis that flags that a request is being processed", "err", err)
+		}
+		delete(p.promises, id)
+	}
 	p.promisesLock.Lock()
 	defer p.promisesLock.Unlock()
 	responded := 0
 	errored := 0
-	for id, promise := range p.promises {
+	for id, msgIDAndPromise := range p.promises {
 		if ctx.Err() != nil {
 			return 0
 		}
-		res, err := p.client.Get(ctx, id).Result()
+		msgKey := MessageKeyFor(p.redisStream, id)
+		res, err := p.client.Get(ctx, msgKey).Result()
 		if err != nil {
-			errSetId := setMinIdInt(&minIdInt, id)
-			if errSetId != nil {
-				log.Error("error setting minId", "err", err)
-				return p.cfg.CheckResultInterval
-			}
 			if !errors.Is(err, redis.Nil) {
 				log.Error("Error reading value in redis", "key", id, "error", err)
+			} else {
+				// The request this producer is waiting for has been past its TTL or is older than current PEL's lower,
+				// so safe to error and stop tracking this promise
+				allowedOldestID := fmt.Sprintf("%d-0", time.Now().Add(-p.cfg.RequestTimeout).UnixMilli())
+				if pelData != nil && pelData.Lower != "" {
+					allowedOldestID = pelData.Lower
+				}
+				if cmpMsgId(msgIDAndPromise.msgID, allowedOldestID) == -1 {
+					msgIDAndPromise.promise.ProduceError(errors.New("error getting response, request has been waiting for too long"))
+					log.Error("error getting response, request has been waiting past its TTL")
+					errored++
+					deletePromise(id)
+				}
 			}
 			continue
 		}
 		var resp Response
 		if err := json.Unmarshal([]byte(res), &resp); err != nil {
-			promise.ProduceError(fmt.Errorf("error unmarshalling: %w", err))
+			msgIDAndPromise.promise.ProduceError(fmt.Errorf("error unmarshalling: %w", err))
 			log.Error("Error unmarshaling", "value", res, "error", err)
 			errored++
 		} else {
-			promise.Produce(resp)
+			msgIDAndPromise.promise.Produce(resp)
 			responded++
 		}
-		delete(p.promises, id)
+		deletePromise(id)
 	}
-	var trimmed int64
-	var trimErr error
-	minId := "+"
-	if minIdInt[0] < math.MaxUint64 {
-		minId = fmt.Sprintf("%d-%d", minIdInt[0], minIdInt[1])
-		trimmed, trimErr = p.client.XTrimMinID(ctx, p.redisStream, minId).Result()
-	} else {
-		trimmed, trimErr = p.client.XTrimMaxLen(ctx, p.redisStream, 0).Result()
+	// XDEL on consumer side already deletes acked messages (mark as deleted) but doesnt claim the memory back, XTRIM helps in claiming this memory in normal conditions
```

Comment:
Depending on how much we prioritize freeing memory early vs lower computation required,
but I think that XTRIM might not be needed here - we could save on some additional complexity here and processing on redis server side.

as per the XDEL [docs](https://redis.io/docs/latest/commands/xdel/): 
> Eventually if all the entries in a macro-node are marked as deleted, the whole node is destroyed and the memory reclaimed.

Reply:
I think we should prioritize freeing memory early even though implications of XDEL would eventually do it any way sometime after all the elements in a macro-node are marked as deleted mainly because- if there is a scenario where one/few consumers fail to delete their ack-ed entries, this will lead to growing of memory indefinitely as ack-ed entries don't show up in PEL and there's really now for us to detect this other than seeing the logs on consumer side. 
Besides I think having a deletion mechanism on both producer and consumer side is a bit safer, I do agree with your point that we don't have to do xtrim as regularly as `CheckResultInterval`, maybe at a 5*`CheckResultInterval` duration sounds reasonable?

Reply:
In order to not miss this, I am including this change with xtrim duration as `5*CheckResultInterval`
tagging @tsahee

Reply:
I agree. My experience so far with redis is that "eventually" might take too long. Also, we want the mechanism to remove very-old entries. However, it is good to know we have another mechanism that clears memory in time.